lib/logical_solver/presolve.cpp

```diff
@@ -0,0 +1,889 @@
+    } else {
+      infeasible = true;
+    }
+  };
+
+  void Presolve::updateStateRowSingletonInequality(
+    int row_index, int col_index
+  ) {
+    // Get the implied bound by divinding the lower bound on the
+    // constraint by the coefficient.
+    int variable_coefficient = problem_matrix_.at(row_index).at(col_index);
+
+    if (variable_coefficient < 0) {
+      // If coefficient negative, upper bound becomes 
+      // lower bound divided by coefficient, and we use floor
+      // to make sure that the bound is met in integer arithmetic.
+      int implied_bound = std::floor(
+        float(lower_bounds_.at(row_index))/float(variable_coefficient)
+      );
+      implied_lower_bounds_.at(col_index) = -implied_upper_bounds_.at(col_index);
+      implied_upper_bounds_.at(col_index) = implied_bound;
+
+    } else {
+      // If coefficient not negative, lower bound becomes lower
+      // bound divided by the coefficient and we use ceil to make
+      // bound is met in integer arithmetic.
+      int implied_bound = std::ceil(
+        float(lower_bounds_.at(row_index))/float(variable_coefficient)
+      );
+      implied_lower_bounds_.at(col_index) = implied_bound;
+    }
+  };
+
+  void Presolve::applyRowSingletonPostsolve(int row_index){
+    // Since row singleton equality leads to a fixed column 
+    // in postsolve the feasible value will already have been
+    // found by applyFixedColPostsolve, so we just have to turn 
+    // the row on on postsolve so that it can be included in the 
+    // constraints checks.
+    if (isRowActivePostsolve(row_index)) {
+      if (checkConstraint(row_index, 1)) {
+        postsolve_active_rows_.at(row_index) = true;
+      }
+    }
+  };
+
+  bool Presolve::checkAreRowsParallel(int row1_index, int row2_index) {
+    if (rows_non_zero_variables_.at(row1_index).size() != rows_non_zero_variables_.at(row2_index).size()) {
+      return false;
+    }
+    int row1_first_non_zero_col = rows_non_zero_variables_.at(row1_index).at(0);
+    int row2_first_non_zero_col = rows_non_zero_variables_.at(row2_index).at(0);
+
+    if (row1_first_non_zero_col != row2_first_non_zero_col) {
+      return false;
+    }
+    double ratio = (double)problem_matrix_.at(row1_index).at(row1_first_non_zero_col)/problem_matrix_.at(row2_index).at(row2_first_non_zero_col);
+
+    for (int j = 1; j < rows_non_zero_variables_.at(row1_index).size(); j++) {
+      int row1_non_zero_col = rows_non_zero_variables_.at(row1_index).at(j);
+      int row2_non_zero_col = rows_non_zero_variables_.at(row2_index).at(j);
+
+      if (row1_non_zero_col != row2_non_zero_col) {
+        return false;
+      }
+
+      double ratio_new = (double)problem_matrix_.at(row1_index).at(row1_non_zero_col)/problem_matrix_.at(row2_index).at(row2_non_zero_col);
+      if (ratio != ratio_new) {return false;}
+      ratio = ratio_new;
+    }
+
+    return true;
+  };
+
+  int Presolve::getParallelRow(int row_index, int start) {
```

Comment:
Feels like you could use optional return types here?

Reply:
Might do if we have time.